Second (control) mempool implementation

viquezclaudio opened the pull request:
- Introduced a second mempool, used for control transactions
- When blocks are produced, first we try to include control txns
- If there is space left, we include regular txns
- This means that control txns have a higher priority over reg txns.

viquezclaudio committed:
Implementation of second mempool

- Introduced a second mempool, used for control transactions
- When blocks are produced, first we try to include control txns
- If there is space left, we include regular txns
- This means that control txns have a higher priority over reg txns.

sisou commented on consensus/src/consensus/mod.rs:
```diff
@@ -42,6 +43,9 @@ impl<N: Network> Clone for ConsensusProxy<N> {
+        if tx.sender_type == AccountType::Staking || tx.recipient_type == AccountType::Staking {
+            return self.network.publish::<ControlTransactionTopic>(tx).await;
+        }
         self.network.publish::<TransactionTopic>(tx).await
```
So control transactions are relayed on both the control-tx topic and also still the regular tx topic?

viquezclaudio replied:
Nop, because there is a return statement in the line where we publish to the ControlTransactionTopic

sisou replied:
Ah, right! 👍

sisou commented on mempool/src/executor.rs:
```diff
@@ -114,3 +114,100 @@ impl<N: Network> Future for MempoolExecutor<N> {
+    // Ongoing verification tasks counter
+    verification_tasks: Arc<AtomicU32>,
+
+    // Reference to the network, to alow for message validation
```
```suggestion
    // Reference to the network, to allow for message validation
```

sisou commented on mempool/src/mempool.rs:
```diff
@@ -89,17 +99,24 @@ impl Mempool {
     /// Once this function is called, the mempool executor is spawned.
     /// The executor will subscribe to the transaction topic from the the network.
-    pub async fn start_executor<N: Network>(&self, network: Arc<N>, monitor: Option<TaskMonitor>) {
+    pub async fn start_executor<N: Network>(
```
Should this function (and its descriptive comments) not be renamed to executor**s** (plural)?

sisou commented on mempool/src/mempool.rs:
```diff
@@ -89,17 +99,24 @@ impl Mempool {
+        &self,
+        network: Arc<N>,
+        monitor: Option<TaskMonitor>,
+        ctrl_monitor: Option<TaskMonitor>,
```
Suggestion: Do not abbreviate `control`  to `ctrl` here, as you are not doing it anywhere else (so far). Should be consistently named `control_monitor`.

sisou commented on mempool/src/mempool.rs:
```diff
@@ -127,6 +144,34 @@ impl Mempool {
+            .await
+            .unwrap();
+
+        let control_mempool_executor = ControlMempoolExecutor::new(
+            Arc::clone(&self.blockchain),
+            Arc::clone(&self.state),
+            Arc::clone(&self.filter),
+            Arc::clone(&network),
+            txn_stream,
+        );
+
+        // Start the executor and obtain its handle
+        let (abort_handle, abort_registration) = AbortHandle::new_pair();
+
+        let future = Abortable::new(control_mempool_executor, abort_registration);
+
+        if let Some(monitor) = ctrl_monitor {
+            tokio::spawn(monitor.instrument(future));
+        } else {
+            tokio::spawn(future);
+        }
+
+        // Set the control mempool executor handle
+        *control_executor_handle = Some(abort_handle);
```
This code is the exact same as in the `start_control_executor_with_txn_stream` function below. Could you call that function here instead?

viquezclaudio replied:
I don't get this, is not exactly the same. As in normal operation, we call the _start_executors_ which would subscribe to the topics and manage the txn topics themselves, whereas the _*_with_txn_stream_ functions are used only for testing purposes and it is the callers responsibility to manage the txn topics..

sisou commented on mempool/src/mempool.rs:
```diff
@@ -162,14 +207,48 @@ impl Mempool {
+        txn_stream: BoxStream<'static, (Transaction, <N as Network>::PubsubId)>,
+        network: Arc<N>,
+    ) {
+        let mut executor_handle = self.control_executor_handle.lock().await;
```
Suggestion to rename. I know it's not necessary, but it will prevent confusion when scanning the function body, which of the two functions (executor vs control_executor) I'm currently in.
```suggestion
        let mut control_executor_handle = self.control_executor_handle.lock().await;
```

sisou replied:
On a similar note: I see a lot of code duplication. Is it maybe possible to generalize the code so that the logic is only written once but called with different params for the regular executor and control_executor?

viquezclaudio replied:
Yes I thought about that, but since the control/regular executor are really distinct entities, then even if we only have a single function that receives a parameter to decide which executor to create, we would still need to instantiate both executors, both handles, etc. so there are not much gains I think

sisou commented on mempool/src/mempool.rs:
```diff
@@ -162,14 +207,48 @@ impl Mempool {
+    /// Stops the mempool executors
     ///
     /// This functions should only be called only after one of the functions to start the executor is called.
     pub async fn stop_executor<N: Network>(&self, network: Arc<N>) {
```
```suggestion
    pub async fn stop_executors<N: Network>(&self, network: Arc<N>) {
```

sisou commented on mempool/src/mempool.rs:
```diff
@@ -89,17 +99,24 @@ impl Mempool {
+        let mut control_executor_handle = self.control_executor_handle.lock().await;
 
-        if executor_handle.is_some() {
+        if executor_handle.is_some() && control_executor_handle.is_some() {
             // If we already have an executor running, don't do anything
             return;
         }
```
This now only returns early, if **both** of the executors are running. If only one is running, both are recreated. Or is that something that 100% cannot happen? Or is recreating an existing executor not a problem?

sisou replied:
And since the control transactions are sent on both topics, do all node types have both mempools? Or is the control mempool only relevant for validator nodes? If so, should the control mempool then be decoupled, so that it only runs in the validator configuration?

viquezclaudio replied:
Both executors are created/terminated in tandem. If for some reason one of them crashes we don't have a way to recreate it (this applies to the current mempool as well, if the executor dies we are not monitoring its health and this would indicate a major problem in our code)
If a node has a mempool, it will have both a control and regular executors.

sisou commented on mempool/src/mempool.rs:
```diff
@@ -178,6 +257,18 @@ impl Mempool {
+        // Stop the control executor
+        control_executor_handle
+            .take()
+            .expect("Expected an control executor handle")
```
```suggestion
            .expect("Expected a control executor handle")
```

sisou commented on mempool/src/mempool.rs:
```diff
@@ -198,6 +289,27 @@ impl Mempool {
+        // Stop the executor
+        handle
+            .take()
+            .expect("Expected an control executor handle")
```
```suggestion
            .expect("Expected a control executor handle")
```

sisou commented on mempool/src/mempool.rs:
```diff
@@ -511,11 +616,77 @@ impl Mempool {
+        let mut mempool_state_upgraded = RwLockUpgradableReadGuard::upgrade(state);
+
+        loop {
+            // Get the hash of the highest paying regular transaction.
```
```suggestion
            // Get the hash of the highest paying control transaction.
```

sisou commented on mempool/src/mempool.rs:
```diff
@@ -558,17 +729,35 @@ impl Mempool {
-        self.state.read().transactions.keys().cloned().collect()
+        let state = self.state.read();
+
+        state
+            .regular_transactions
+            .transactions
+            .keys()
+            .cloned()
+            .chain(state.control_transactions.transactions.keys().cloned())
+            .collect()
```
This returns regular transactions before control transactions. But control transactions are supposed to be higher importance. Should they not be listed first?

viquezclaudio replied:
Well I think that prioritization matters only during block production, since we want to include control transaction first.
Do you think it also matters for this function? Basically to get the control transactions hashes first as well for RPC purposes

sisou replied:
It would be consistent. And would also show RPC clients that those transactions are higher priority. If some where to write a fee estimator, or a confirmation estimation, that can make sense, as the control txs are included first, and the client would not need to sort them again.

But maybe that is just premature optimization :shrug:

sisou commented on mempool/src/mempool.rs:
```diff
@@ -558,17 +729,35 @@ impl Mempool {
+            .values()
+            .cloned()
+            .chain(state.control_transactions.transactions.values().cloned())
+            .collect()
```
This returns regular transactions before control transactions. But control transactions are supposed to be higher importance. Should they not be listed first?

viquezclaudio replied:
Same as before. Does this matter for RPC purposes?

sisou commented on mempool/src/mempool.rs:
```diff
@@ -627,21 +804,62 @@ pub(crate) struct MempoolState {
+        self.transactions.is_empty()
+    }
+
+    //This function is used to remove the transactions that are no longer valid at a given block height
```
```suggestion
    // This function is used to remove the transactions that are no longer valid at a given block height
```

sisou commented on mempool/src/mempool.rs:
```diff
@@ -664,11 +882,89 @@ impl MempoolState {
         self.tx_counter += 1;
 
         self.oldest_transactions
             .push(tx_hash.clone(), Reverse(tx.validity_start_height));
```
Clippy warning here: redundant clone

sisou commented on mempool/src/mempool.rs:
```diff
@@ -664,11 +882,89 @@ impl MempoolState {
+        self.total_size += tx.serialized_size();
+
+        true
+    }
```
I don't understand how this comment describes what's happening. The `total_size` is increased, but I do not see any removal of a cheapest one, or a check for too many txs?

viquezclaudio replied:
outdated comment, removal of cheapest ones is done outside this function

sisou commented on mempool/src/mempool.rs:
```diff
@@ -664,11 +882,89 @@ impl MempoolState {
+
+    pub fn get(&self, hash: &Blake2bHash) -> Option<&Transaction> {
+        if let Some(transaction) = self.regular_transactions.get(hash) {
+            return Some(transaction);
```
Clippy warnings here: unneeded `return` statements

sisou commented on mempool/src/mempool.rs:
```diff
@@ -664,11 +882,89 @@ impl MempoolState {
+            return false;
+        }
+
+        // If we are adding a stacking transaction we insert it into the control txns container
```
This comment describes **what** the code does. Which I can see in the code. It should instead describe **why** it does it. It should explain that staking txns _are_ control txns (which is not explained anywhere else yet).
```suggestion
        // Staking txns are control txns
```

viquezclaudio replied:
Yes, for now we decided to consider staking transactions = control transactions. 
But I think the is an ongoing discussion..

sisou commented on mempool/src/mempool.rs:
```diff
@@ -664,11 +882,89 @@ impl MempoolState {
+
+        // If we are adding a stacking transaction we insert it into the control txns container
+        if tx.sender_type == AccountType::Staking || tx.recipient_type == AccountType::Staking {
+            self.control_transactions.insert(&tx);
```
Clippy warning about creating an immediately dereferenced reference

sisou commented on mempool/src/mempool.rs:
```diff
@@ -664,11 +882,89 @@ impl MempoolState {
+            self.regular_transactions.insert(&tx);
+        }
+
+        //Update the per sender state
```
```suggestion
        // Update the per sender state
```

sisou commented on mempool/src/mempool.rs:
```diff
@@ -754,39 +1053,52 @@ impl MempoolState {
+            self.metrics.note_evicted(reason);
 
-        self.remove_from_staking_state(&tx);
+            return Some(tx);
```
Clippy warning: unneeded `return` statements

sisou commented on mempool/src/mempool.rs:
```diff
@@ -754,39 +1053,52 @@ impl MempoolState {
+            let sender_state = self.state_by_sender.get_mut(&tx.sender).unwrap();
 
-        self.total_size -= tx.serialized_size();
+            sender_state.total -= tx.total_value();
+            sender_state.txns.remove(tx_hash);
 
-        #[cfg(feature = "metrics")]
-        self.metrics.note_evicted(reason);
+            if sender_state.txns.is_empty() {
+                self.state_by_sender.remove(&tx.sender);
+            }
 
-        Some(tx)
+            self.remove_from_staking_state(&tx);
+
+            #[cfg(feature = "metrics")]
+            self.metrics.note_evicted(reason);
+
+            return Some(tx);
```
It looks to me like the code that gets run when a transaction is removed is the **same** for both regular and control transactions, because the `sender_state` is global across both types?

In that case, the code should just be declared once and run if the deleted tx existed in either tx-type bucket, e.g. by combining the condition:
```rust
if let Some(tx) = self.regular_transactions.delete(tx_hash) || self.control_transactions.delete(tx_hash) {
```
Would that be possible?

sisou replied:
Also, Clippy warning about unneeded `return` statement

viquezclaudio replied:
I wanted to do that, but that is not permitted in the language.
They are planning to introduce some chain "let" assignments:
https://github.com/rust-lang/rust/issues/53667

But it is still forbidden with the || operator:
https://github.com/rust-lang/rust/pull/94754

sisou commented on mempool/src/mempool.rs:
```diff
@@ -754,39 +1053,52 @@ impl MempoolState {
-                self.worst_transactions.remove(tx_hash);
-                self.oldest_transactions.remove(tx_hash);
-                if let Some(tx) = self.transactions.remove(tx_hash) {
+                if let Some(tx) = self.regular_transactions.delete(tx_hash) {
+                    self.remove_from_staking_state(&tx);
+                }
+
+                if let Some(tx) = self.control_transactions.delete(tx_hash) {
                     self.remove_from_staking_state(&tx);
                 }
```
Similar to above, can this be combined?
```rust
if let Some(tx) = self.regular_transactions.delete(tx_hash) || self.control_transactions.delete(tx_hash) {
    self.remove_from_staking_state(&tx);
}
```

viquezclaudio replied:
Same as above, forbidden. 
Probably the reason is: what happens if both regular_transactions and control_transactions return an object?
if they both return an object, tx would be assigned to what? to the object returned by the regular_transactions or the one returned by control_transactions?

sisou replied:
In Javascript, the functions are run after each other. If the first call returns a truthy value, the second function is not called anymore. That's what I would expect. And in this case it makes sense, because only one tx bucket can contain the tx_hash in this implementation.

nibhar replied:
Just saw this.

`||` does not work, but `or_else(...)` should.

```suggestion
                if let Some(tx) = self.regular_transactions.delete(tx_hash)
                    .or_else(|| self.control_transactions.delete(tx_hash))
                { 
                    self.remove_from_staking_state(&tx);
                }
```

sisou commented on validator/src/micro.rs:
```diff
@@ -240,9 +240,20 @@ impl<TValidatorNetwork: ValidatorNetwork + 'static> NextProduceMicroBlockEvent<T
         );
 
-        let transactions = self
+        //First we try to fill the block with control transactions
```
```suggestion
        // First we try to fill the block with control transactions
```

sisou commented on validator/src/validator.rs:
```diff
@@ -211,6 +213,8 @@ impl<TNetwork: Network, TValidatorNetwork: ValidatorNetwork>
             #[cfg(feature = "metrics")]
             mempool_monitor: TaskMonitor::new(),
+            #[cfg(feature = "metrics")]
+            ctrl_mempool_monitor: TaskMonitor::new(),
```
Suggestion to rename, to be consistent across the codebase.
```suggestion
            control_mempool_monitor: TaskMonitor::new(),
```

viquezclaudio commented:
@sisou any more comments?

viquezclaudio committed:
Include control mempool config option

Create a new config parameter to limit the amount of transactions
that can be kept in the control mempool

nibhar commented on mempool/src/executor.rs:
```diff
@@ -114,3 +114,100 @@ impl<N: Network> Future for MempoolExecutor<N> {
+            filter,
+            network,
+            network_id: Arc::new(blockchain.read().network_id),
+            verification_tasks: Arc::new(AtomicU32::new(0)),
```
More of a question than an actual comment:

So each mempool has their ow transaction verification task limit. That doubles the amount of verification tasks from what it was previously. Is that intended?

viquezclaudio replied:
Yes, I think we never really hit that limit under normal operation, should I set it to half of it, so we maintain the same total amount of verification tasks as before?

nibhar commented on mempool/src/mempool.rs:
```diff
@@ -54,27 +66,28 @@ pub struct Mempool {
 
     /// Creates a new mempool
     pub fn new(blockchain: Arc<RwLock<Blockchain>>, config: MempoolConfig) -> Self {
         let state = MempoolState {
```
Maybe add a `MempoolState::new(size_limit: usize) -> MempoolState { ... }`

nibhar commented on mempool/src/mempool.rs:
```diff
@@ -511,11 +619,77 @@ impl Mempool {
+    /// Returns a vector with accepted control transactions from the mempool.
+    ///
+    /// Returns the highest fee per byte up to max_bytes transactions and removes them from the mempool
+    pub fn get_control_transactions_for_block(
```
I moved the logic for the expired transactions into the `MempoolState`. I did not do it for `get_transactions_for_block` because I am unsure wether or not more fine grained control over the byte size in particular is required.
However I believe even then a function like:
```
impl MempoolState {
    pub fn get_transactions_for_block(&self, max_bytes: (usize, usize)) -> Vec<Transactions>) {
        ...
    }
}
```
might be the way to go here. One thing in particular that I see is that for both the regular and the control transactions a `self.state.upgradable_read()` and a subsequent upgrade of the lock are required. Moving the logic to the `MempoolState` would remove that. The code is also pretty much identical, aside from which field to access. Abstracting what transactions shall be retrieved might also be a good idea.

If that is not desirable, I probably would still move both these functions in the more lightweight `MempoolState` and just call them from here after taking the lock.

viquezclaudio replied:
if we move to the `pub fn get_transactions_for_block(&self, max_bytes: (usize, usize))` format, how would you specify what we are currently doing? i.e.: we know how much available bytes we have in the block, and we want to fill the block with control txns and if there is space left, fill the remaining with regular txns?

viquezclaudio replied:
I still agree about the other benefits, the code its pretty much identical and it would be a good idea to move it to the `MempoolState`

nibhar reviewed (commented):
I created #929 to some smaller refactoring and to move some code pieces to their own files. Feel free to look through it and cherry pick if necessary.

Other than that I have little to add to what @sisou already provided. I do agree that control transactions should always come first, as there is no reason to do it the other way around and it does not introduce any complexity as far as I can tell.

nibhar committed:
Second mempool, small refactoring (#929)

* Move some structs to their own modules

All logic remains unchanged, this commit only moves code around.

* Only use a single MempoolExecutor by using generic Topic

* Add private fn start_executor and reuse where applicable

* Move expired_transactions function to MempoolState

viquezclaudio committed:
Small fixes to mempool code

- Small warning/compilation fixes
- Return control txns first, regular txns second
- Share the max number of verif tasks among multiple executors
- create constructor for mempool state
